rust/rsc/src/database.rs

```diff
@@ -645,37 +607,45 @@ pub async fn total_blob_size<T: ConnectionTrait>(db: &T) -> Result<Option<TotalB
 // ----------            Update            ----------
 
 // ----------            Delete            ----------
-pub async fn delete_blobs_by_ids<T: ConnectionTrait>(db: &T, ids: Vec<Uuid>) -> Result<u64, DbErr> {
-    if ids.len() == 0 {
-        return Ok(0);
-    }
-
-    let mut affected = 0;
-
-    let chunked: Vec<Vec<Uuid>> = ids
-        .into_iter()
-        .chunks((MAX_SQLX_PARAMS / 1).into())
-        .into_iter()
-        .map(|chunk| chunk.collect())
-        .collect();
+#[derive(Clone, Debug, FromQueryResult)]
+pub struct DeletedBlob {
+    pub store_id: Uuid,
+    pub key: String,
+}
 
-    for chunk in chunked {
-        let result = Blob::delete_many()
-            .filter(
-                entity::blob::Column::Id.in_subquery(
-                    migration::Query::select()
-                        .column(migration::Asterisk)
-                        .from_values(chunk, migration::Alias::new("foo"))
-                        .take(),
-                ),
+// Deletes blobs from the database that are unreferenced and have surpassed the allocated grace
+// period to be referenced.
+//
+// For new blobs this allows the client to create several blobs and then reference them all at
+// once. Existing blobs whose job was just evicted will likely be well past the grace period and
+// thus quickly evicted themselves.
+pub async fn delete_unreferenced_blobs<T: ConnectionTrait>(
+    db: &T,
+    ttl: NaiveDateTime,
+    chunk: u32,
+) -> Result<Vec<DeletedBlob>, DbErr> {
+    DeletedBlob::find_by_statement(Statement::from_sql_and_values(
+        DbBackend::Postgres,
+        r#"
+            WITH
+            eligible_blob_ids as (
+                SELECT id FROM blob
+                WHERE updated_at <= $1
+                EXCEPT (
+                    SELECT blob_id FROM output_file
+                    UNION SELECT stdout_blob_id FROM job
+                    UNION SELECT stderr_blob_id FROM job
+                )
+                LIMIT $2
             )
-            .exec(db)
-            .await?;
-
-        affected += result.rows_affected;
-    }
-
-    Ok(affected)
+            DELETE from blob b
+            WHERE b.id IN (SELECT id FROM eligible_blob_ids)
```

Comment:
Is it not possible to just do `IN (eligible_blob_ids.id)` or something like that? I'm surprised to see a select is needed again.

Reply:
Nope :( I tried that. As fair as I could tell this didn't actually effect the query plan since `WITH` is kind of just writing an alias for a subquery